mirage/client/dns_client_mirage.ml

```diff
@@ -1,12 +1,105 @@
+    in
+    match String.split_on_char ':' str with
+    | [] -> Fmt.failwith "Invalid fingerprint %S" str
+    | [ fp ] -> `SHA256, fp
+    | hash :: rest -> (
+        match hash_of_string (String.lowercase_ascii hash) with
+        | Some hash -> hash, String.concat "" rest
+        | None -> Fmt.failwith "Invalid hash algorithm: %S" hash)
+  in
+  let fp =
+    try Hex.to_cstruct (`Hex fp)
+    with _ -> Fmt.failwith "Invalid hex fingerprint value: %S" fp
+  in
+  hash, fp
+
 let src = Logs.Src.create "dns_client_mirage" ~doc:"effectful DNS client layer"
 module Log = (val Logs.src_log src : Logs.LOG)
 
 module IM = Map.Make(Int)
 
+module type S = sig
+  module Transport : Dns_client.S
+    with type io_addr = [ `Plaintext of Ipaddr.t * int | `Tls of Tls.Config.client * Ipaddr.t * int ]
+     and type +'a io = 'a Lwt.t
+
+  include module type of Dns_client.Make(Transport)
+
+  val authenticators_of_strings : string list -> X509.Authenticator.t Ipaddr.Map.t
+  val nameserver_of_string : X509.Authenticator.t Ipaddr.Map.t -> string -> Transport.io_addr
+  val connect_device :
+    ?protocol:string ->
+    ?authenticators:string list ->
+    ?nameservers:string list ->
+    ?timeout:int64 ->
+    Transport.stack -> t Lwt.t
+end
+
 module Make (R : Mirage_random.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (S : Tcpip.Stack.V4V6) = struct
 
+  let authenticator_of_string str =
+    let time () = Some (Ptime.v (P.now_d_ps ())) in
+    match String.split_on_char ':' str with
+    | nameserver :: "key" :: tls_key_fingerprint ->
+      let ipaddr = Ipaddr.of_string_exn nameserver in
+      let tls_key_fingerprint = String.concat ":" tls_key_fingerprint in
+      let hash, fingerprint = of_fp tls_key_fingerprint in
+      ipaddr, X509.Authenticator.server_key_fingerprint ~time ~hash ~fingerprint
+    | nameserver :: "cert" :: tls_cert_fingerprint ->
+      let ipaddr = Ipaddr.of_string_exn nameserver in
+      let tls_cert_fingerprint = String.concat ":" tls_cert_fingerprint in
+      let hash, fingerprint = of_fp tls_cert_fingerprint in
+      ipaddr, X509.Authenticator.server_cert_fingerprint ~time ~hash ~fingerprint
+    | nameserver :: "trust-anchor" :: certs ->
+      let ipaddr = Ipaddr.of_string_exn nameserver in
+      let certs = List.map Base64.decode certs in
+      let certs = List.map (Result.map Cstruct.of_string) certs in
+      let certs = List.map
+        (fun cert -> Result.bind cert X509.Certificate.decode_der)
+        certs in
+      let certs = List.filter_map Result.to_option certs in
+      ipaddr, X509.Authenticator.chain_of_trust ~time certs
+    | [ nameserver; "none" ] ->
+      let ipaddr = Ipaddr.of_string_exn nameserver in
+      ipaddr, none
+    | _ -> Fmt.failwith "Invalid TLS authenticator: %S" str
+
+  let authenticators_of_strings lst =
+    let lst = List.map authenticator_of_string lst in
+    List.fold_left (fun a (k, v) -> Ipaddr.Map.add k v a) Ipaddr.Map.empty lst
```

Comment:
thanks, but that code should be in x509, not in dns.

Reply:
If you agree with the format (already used by `ocaml-git` for the HTTPS authenticator), I can propose it on `x509` yes.